pkg/logger/ecs_error_log.go

```diff
@@ -0,0 +1,175 @@
+package logger
+
+import (
+	"fmt"
+	"go.uber.org/zap/zapcore"
+	"os"
+	"path/filepath"
+	"strconv"
+	"time"
+)
+
+var ErrorFilePath = os.Getenv("STATUS_MESSAGE_FILE_PATH")
+
+var errorLogMaxAge = os.Getenv("STATUS_MESSAGE_TTL")
+var errorLogMaxAgeInHours, _ = strconv.Atoi(errorLogMaxAge)
+
+var size = os.Getenv("STATUS_MESSAGE_MAX_BYTE_LENGTH")
+var errorFileSize, _ = strconv.Atoi(size)
+
+var queue []zapcore.Entry
+var errorLogChannel = make(chan zapcore.Entry, 10)
+var ticker = time.NewTicker(10 * time.Minute)
+var errorFile *os.File
+
+// ECSErrorReporter func help select the different case between either write error log into file or
+// rotate and delete old logs in file based on TTL
+func ECSErrorReporter() {
+	defer ticker.Stop()
+	for {
+		select {
+		// when new error add in the channel, this case will write it into file
+		case e := <-errorLogChannel:
+			_, _ = WriteError(e, ErrorFilePath)
+		// this case will run every x mins (x is configurable) to delete the error which already expired, error log max age
+		// depends on STATUS_MESSAGE_TTL
+		case <-ticker.C:
+			currentTime := time.Now()
+			file, _ := os.OpenFile(ErrorFilePath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC|os.O_APPEND, 0777)
+			_ = file.Truncate(0)
+			errorNum := len(queue)
+			for i := 0; i < errorNum; i++ {
+				e := queue[0]
+				if e.Time.Add(time.Duration(errorLogMaxAgeInHours) * time.Hour).After(currentTime) {
+					_, _ = file.Write([]byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+						e.Time, e.Level, e.Caller, e.Message)))
+				} else {
+					queue = queue[1:]
+				}
+			}
+			_ = file.Close()
+		}
+	}
+}
+
+// WriteError func could write the new error log message into file
+func WriteError(newErrorLog zapcore.Entry, errorLogfile string) (n int, err error) {
+	var byteNum int
+	if errorFile == nil {
+		if err = openExistingOrNewErrorFile(); err != nil {
+			return 0, err
+		}
+	}
+	err = errorFile.Truncate(0)
+	if err != nil {
+		return byteNum, err
+	}
+	// iterate errors in queue, if the old error's message same as the new error's message, skip it,
+	// otherwise write it into the file and queue
+	var errorNum = len(queue)
+	for i := 0; i < errorNum; i++ {
+		e := queue[0]
+		if e.Message != newErrorLog.Message {
+			n, err = errorFile.Write([]byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+				e.Time, e.Level, e.Caller, e.Message)))
+			if err != nil {
+				return byteNum, err
+			}
+			byteNum += n
+			queue = append(queue, e)
+		}
+		queue = queue[1:]
+	}
+	queue = append(queue, newErrorLog)
```

Comment:
Based on the design, you are supposed to only add the new error log to the queue if it is not already in the queue. That way there are no duplicates in the queue or the file.